Parse web extension manifest for declarativeNetRequest rulesets

b-weinstein opened the pull request:
#### 1093d6012bcfd5c17dbd4710264729b559612699
<pre>
Parse web extension manifest for declarativeNetRequest rulesets
<a href="https://bugs.webkit.org/show_bug.cgi?id=264836">https://bugs.webkit.org/show_bug.cgi?id=264836</a>
<a href="https://rdar.apple.com/118415584">rdar://118415584</a>

Reviewed by Timothy Hatcher.

This patch hooks up basic manifest parsing for declarativeNetRequest rule sets.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.mm:
(-[_WKWebExtension hasDeclarativeNetRequestRules]): Mostly added for testing at this point. Returns if there were any correctly parsed rulesets.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::errors): Call parseDeclarativeNetRequestRulesetDictionary.
(WebKit::WebExtension::declarativeNetRequestRulesets): Call populateDeclarativeNetRequestPropertiesIfNeeded and return the vector of rulesets.
(WebKit::WebExtension::parseDeclarativeNetRequestRulesetDictionary): Takes in a dictionary and returns an optional DeclarativeNetRequestRulesetData. This method
does argument type checking and makes sure the string arguments aren&apos;t nil or empty.
(WebKit::WebExtension::populateDeclarativeNetRequestPropertiesIfNeeded): Read the manifest and turn the JSON into a vector of DeclarativeNetRequestRulesetData. This
PR makes sure the extension doesn&apos;t specify too many rulesets, too many enabled rulesets, or multiple rulesets with the same id.
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
(TestWebKitAPI::TEST): Add some tests for the manifest parsing.

Canonical link: <a href="https://commits.webkit.org/270751@main">https://commits.webkit.org/270751@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8fe2289c69d4b7a4f308a2ce650e64f9e58a7846

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26305 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4914 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27563 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28400 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24074 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26638 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6699 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2325 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24069 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26561 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3757 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/22615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28979 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/3348 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23583 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29646 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/23970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/23981 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/27537 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3387 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/1590 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4811 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6321 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3864 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3711 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1601,6 +1612,128 @@ static bool parseCommandShortcut(const String& shortcut, OptionSet<ModifierFlags
+        *error = createError(WebExtension::Error::InvalidDeclarativeNetRequest, WEB_UI_STRING("Empty ruleset id.", "_WKWebExtensionErrorInvalidDeclarativeNetRequestEntry description for empty ruleset id"));
+        return std::nullopt;
+    }
+
```
According to mdn, id's also cannot start with `_`
https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/declarative_net_request

b-weinstein replied:
That appears to not be something we enforce currently in Safari. I'm happy to file a bug + add a FIXME about it and add it in a follow-up.

b-weinstein replied:
rdar://118435154

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1601,6 +1612,128 @@ static bool parseCommandShortcut(const String& shortcut, OptionSet<ModifierFlags
+
+    m_parsedManifestDeclarativeNetRequestRulesets = true;
+
+    // Documentation: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/declarative_net_request
```
`// Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/manifest.json/declarative_net_request`

xeenon replied:
Yeah drop the en-US so it auto negotiate the language of the viewer.

b-weinstein replied:
Neat!

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -782,6 +792,7 @@ static _WKWebExtensionError toAPI(WebExtension::Error error)
     populateContentSecurityPolicyStringsIfNeeded();
     populateWebAccessibleResourcesIfNeeded();
     populateCommandsIfNeeded();
+    populateDeclarativeNetRequestPropertiesIfNeeded();
```
can update the `// FIXME: <https://webkit.org/b/246493>` above since you've added the dnr errors

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1601,6 +1612,128 @@ static bool parseCommandShortcut(const String& shortcut, OptionSet<ModifierFlags
+
+    // Documentation: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/declarative_net_request
+
+    if (!supportedPermissions().contains(_WKWebExtensionPermissionDeclarativeNetRequest) && !supportedPermissions().contains(_WKWebExtensionPermissionDeclarativeNetRequestWithHostAccess)) {
```
not sure i understand what this check is for. The error message below makes me think it's the case where the manifest has a `declarativeNetRequestManifestKey` but doesn't claim the `declarativeNetRequest` permission?

b-weinstein replied:
That is correct.

b-weinstein replied:
Which appears to be documented in:
https://developer.chrome.com/docs/extensions/reference/declarativeNetRequest/

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1601,6 +1612,128 @@ static bool parseCommandShortcut(const String& shortcut, OptionSet<ModifierFlags
+    };
+
+    NSString *exceptionString;
+    BOOL isRulesetDictionaryValid = validateDictionary(rulesetDictionary, nil, requiredKeysInRulesetDictionary, keyToExpectedValueTypeInRulesetDictionary, &exceptionString);
```
`bool`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1601,6 +1612,128 @@ static bool parseCommandShortcut(const String& shortcut, OptionSet<ModifierFlags
+        (bool)objectForKey<NSNumber>(rulesetDictionary, declarativeNetRequestRuleEnabledManifestKey).boolValue,
+
+        // jsonPath
+        jsonPath
```
The comments don't add much here. This could be three lines.

b-weinstein replied:
Yeah I realized that once I was writing them.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1601,6 +1612,128 @@ static bool parseCommandShortcut(const String& shortcut, OptionSet<ModifierFlags
+        jsonPath
+    };
+
+    return std::optional { rulesetData };
```
It would be better to have these together, to avoid the copy.

```
return { DeclarativeNetRequestRulesetData {
    rulesetID,
    (bool)objectForKey<NSNumber>(rulesetDictionary, declarativeNetRequestRuleEnabledManifestKey).boolValue,
    jsonPath
} };
```

Our maybe `return { WTFMove(rulesetData) }` will work (the compiler will complain otherwise).

b-weinstein replied:
I adopted `return { WTFMove(rulesetData) }`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1601,6 +1612,128 @@ static bool parseCommandShortcut(const String& shortcut, OptionSet<ModifierFlags
+
+    NSString *rulesetID = objectForKey<NSString>(rulesetDictionary, declarativeNetRequestRulesetIDManifestKey);
+    if (!rulesetID.length) {
+        *error = createError(WebExtension::Error::InvalidDeclarativeNetRequest, WEB_UI_STRING("Empty ruleset id.", "_WKWebExtensionErrorInvalidDeclarativeNetRequestEntry description for empty ruleset id"));
```
All these error strings should mention `declarativeNetRequest` or `declarative_net_request` depending on where the error is (in this case the manifest, so underscore version).

Something like:

```
WEB_UI_STRING("Missing or empty `id` in `declarative_net_request` manifest entry.", …)
```

Otherwise there will not be enough context for the developer who will see these, in Safari at least in Settings.

b-weinstein replied:
Will do!

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1601,6 +1612,128 @@ static bool parseCommandShortcut(const String& shortcut, OptionSet<ModifierFlags
+    bool recordedTooManyRulesetsManifestError = false;
+    HashSet<String> seenRulesetIDs;
+    for (NSDictionary *rulesetDictionary in declarativeNetRequestRulesets) {
+        if (rulesetCount == maximumNumberOfDeclarativeNetRequestStaticRulesets)
```
`>=` just in case, or was this suppose to use `recordedTooManyRulesetsManifestError`?

b-weinstein replied:
I think the logic is that if we get to exactly `maximumNumberOfDeclarativeNetRequestStaticRulesets`, then we will always hit the early continue and we will never add any more, but I can change it to `>=` to be more clear.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1601,6 +1612,128 @@ static bool parseCommandShortcut(const String& shortcut, OptionSet<ModifierFlags
+        NSError *error;
+        auto optionalRuleset = parseDeclarativeNetRequestRulesetDictionary(rulesetDictionary, &error);
+        if (!optionalRuleset) {
+            recordError(error);
```
Should be:

```
recordError(createError(Error::InvalidDeclarativeNetRequest, nil, error))
```

That will use this and include the parse error:

```
WEB_UI_STRING("Unable to parse `declarativeNetRequest` rules: %@", "WKWebExtensionErrorInvalidDeclarativeNetRequest description, because of a JSON error")
```

b-weinstein replied:
Nice!

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1601,6 +1612,128 @@ static bool parseCommandShortcut(const String& shortcut, OptionSet<ModifierFlags
+
+        if (ruleset.enabled && ++enabledRulesetCount > maximumNumberOfDeclarativeNetRequestEnabledStaticRulesets && !recordedTooManyRulesetsManifestError) {
+            recordError(createError(Error::InvalidDeclarativeNetRequest, WEB_UI_FORMAT_STRING("Exceeded maxmimum number of enabled static rulesets. The first %lu will be applied, the remaining will be ignored.", "_WKWebExtensionErrorInvalidDeclarativeNetRequestEntry description for too many enabled static rulesets", maximumNumberOfDeclarativeNetRequestEnabledStaticRulesets)));
+            recordedTooManyRulesetsManifestError = true;
```
Is this used?

b-weinstein replied:
Yeah - it's used at the end of line 1718, so we only generate one of these errors.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1601,6 +1612,128 @@ static bool parseCommandShortcut(const String& shortcut, OptionSet<ModifierFlags
+        }
+
+        seenRulesetIDs.add(ruleset.rulesetID);
+        rulesetCount++;
```
`++rulesetCount;`

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.h:
```diff
@@ -279,10 +288,14 @@ class WebExtension : public API::ObjectImpl<API::Object::Type::WebExtension>, pu
     void populateCommandsIfNeeded();
+    void populateDeclarativeNetRequestPropertiesIfNeeded();
+
+    std::optional<WebExtension::DeclarativeNetRequestRulesetData> parseDeclarativeNetRequestRulesetDictionary(NSDictionary *rulesetDictionary, NSError **);
```
Can drop `rulesetDictionary` param name.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
```diff
@@ -246,6 +246,9 @@ WK_CLASS_AVAILABLE(macos(13.3), ios(16.4))
 @property (nonatomic, readonly) BOOL hasCommands;
 
+/*! @abstract A boolean value indicating whether the extension includes declarativeNetRequest rules used for content blocking. */
+@property (nonatomic, readonly) BOOL hasDeclarativeNetRequestRules;
```
I would like to avoid `declarativeNetRequest` in our API. The extension API name is a bit on the nose, and our API should be more generic to be able to morph and adapt as the extension ecosystem evolves.

Maybe:

```
/*! @abstract A boolean value indicating whether the extension includes rules used for content modification or blocking. */
@property (nonatomic, readonly) BOOL hasContentModificationRules;
```

This would allow it to encompass any declarative style rules, since that might be under a new name based on talks in the WECG.

b-weinstein replied:
Sounds good!

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.mm:
```diff
@@ -275,6 +275,11 @@ - (BOOL)hasCommands
 
+- (BOOL)hasDeclarativeNetRequestRules
+{
+    return !_webExtension->declarativeNetRequestRulesets().isEmpty();
```
This would be best to have a method on `WebKit::WebExtension` to call. While this is simple now, it might get more complex in the future and these API wrappers should ideally have almost no logic.

It can just be an inline function in WebExtension.h that does `isEmpty()` for now.

b-weinstein replied:
Got it! So then the extra logic will go into WebExtensionCocoa.mm instead of _WKWebExtension.

b-weinstein committed:
Parse web extension manifest for declarativeNetRequest rulesets
https://bugs.webkit.org/show_bug.cgi?id=264836
rdar://118415584

Reviewed by Timothy Hatcher.

This patch hooks up basic manifest parsing for declarativeNetRequest rule sets.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.mm:
(-[_WKWebExtension hasDeclarativeNetRequestRules]): Mostly added for testing at this point. Returns if there were any correctly parsed rulesets.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::errors): Call parseDeclarativeNetRequestRulesetDictionary.
(WebKit::WebExtension::declarativeNetRequestRulesets): Call populateDeclarativeNetRequestPropertiesIfNeeded and return the vector of rulesets.
(WebKit::WebExtension::parseDeclarativeNetRequestRulesetDictionary): Takes in a dictionary and returns an optional DeclarativeNetRequestRulesetData. This method
does argument type checking and makes sure the string arguments aren't nil or empty.
(WebKit::WebExtension::populateDeclarativeNetRequestPropertiesIfNeeded): Read the manifest and turn the JSON into a vector of DeclarativeNetRequestRulesetData. This
PR makes sure the extension doesn't specify too many rulesets, too many enabled rulesets, or multiple rulesets with the same id.
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
(TestWebKitAPI::TEST): Add some tests for the manifest parsing.

Canonical link: https://commits.webkit.org/270751@main
